app/test_server.py

```diff
@@ -20,17 +21,47 @@ def test_small_test():
 @pytest.mark.xfail(not TRAVIS_CI, reason=REASON)
 def test_invalid_url_api_call():
     response = requests.get('http://localhost:7001/api/v1/search/invalid_url')
-    assert response.json()['Status Code'] == 404
+    # assert response.status_code == 404  # workaround for #332
+    if response.status_code != 404:
+        raise AssertionError(response.status_code)
 
 
 def make_engine_api_call(engine_name):
+    # test that URLs without parameters are rejected as invalid
     url = 'http://localhost:7001/api/v1/search/' + engine_name
-    assert requests.get(url).json()['Status Code'] == 400, engine_name
+    # assert requests.get(url).status_code == 400  # workaround for #332
+    if requests.get(url).status_code != 400:
+        raise AssertionError(engine_name)
+    # These engines need to be fixed so that they pass this test
+    if engine_name in ('dailymotion', 'exalead', 'mojeek', 'yandex'):
+        return
+    # add query parameters to the  URL
+    response = requests.get(url + '?query=fossaisa&num=3')
+    # assert response.status_code == 200, engine_name  # workaround for #332
+    if response.status_code != 200:
+        raise AssertionError(engine_name)
+    if not DEEPER_TESTS:
+        return
+    links = response.json()
+    if isinstance(links, dict):
+        # assert links.get('status_code', 200) == 200  # workaround for #332
+        if links.get('status_code', 200) != 200:
+            raise AssertionError(engine_name)
+    # These engines run in search_without_count() mode so we do not test length
+    if engine_name not in ('quora', 'youtube'):
+        # assert len(links) == 3, '{}: {}'.format(engine_name, len(links))
+        if len(links) != 3:
+            raise AssertionError(engine_name)
+    for link in links:
+        # assert 'link' in link, engine_name   # workaround for #332
+        if 'link' not in link:
```

Comment:
Following two if cases are redundant. It can be sorted by using something like `if ('link' not in link) or ('title' not in link)`.

Reply:
Six of one, half a dozen of the other.  I prefer the original formulation because it is evident to the reader of the test report which element is missing.

Reply:
Okay. In that case, I would wait for the reviews from other members. We need to remove redundancy from the code from all the places if agreed on.

Reply:
In production code, redundancy should be avoided.

In testing code, we should optimize the experience of contributors.  Fast-to-fix is our objective. When a test fails, we want contributors to rapidly find and fix the offending issue.  Focusing them on a single root cause will lighten their cognitive load.

Reply:
Agreed. With this perspective, the changes are well defined to help and are developer friendly.

Reply:
Given that, can you please provide your approval or let me know what is still missing.